---
title: "Differential expression results - deseq2"
author:
  # - Jane Doe^[Institution Two, jane@example.org]      # add report authors (uncomment if using)
  # - John Doe^[Institution One, john@example.org]      # add a second report author (uncomment if using)
date: "Date: `r base::format(Sys.time(), '%d/%m/%Y')`"
always_allow_html: true
output:
  bookdown::html_document2:
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: true
    number_sections: false
    self_contained: yes
    css: ../esr-styles.css
    highlight: pygments
    includes:
      after_body: ../esr_footer.html
  editor_options: 
    chunk_output_type: console
---

<img src="../images/ESR_logo.svg" style="position:absolute;top:19px;right:40px;width:23%;" />

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, error = FALSE, message = FALSE, fig.align = "center")
```

```{r}
# load libraries
library(dplyr)
library(limma)
library(DESeq2)
library(janitor)
library(edgeR)
library(DT)
library(apeglm)
library(plotly)
library(heatmaply)
library(gtools)
library(textshape)
library(tidyr)

# for icons at bottom of doc
htmltools::tagList(rmarkdown::html_dependency_font_awesome())

# read in yaml config file
config <- yaml::yaml.load_file("../config/config.yaml")

# read in metadata
metadata <- utils::read.csv(base::file.path(config$metadata))

# read in deseq differential expression results
deseq_diff_expr_results <- utils::read.table("./diff_expr_results/all_deseq_diff_expr_results.tsv", header = TRUE)
```

## Info

General info:

- Differential expression analysis using [deseq2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8).
- Some values have been rounded

Datasets analysed:

```{r, results = "asis"}
# print the datasets the user has chosen to analyse for this document
base::cat(base::paste0(" - transcript rnaseq: ", config$transcript_rnaseq, "\n",
                       " - gene rnaseq: ", config$gene_rnaseq, "\n"))
```

Treatment comparisons: 

```{r, results = "asis"}
# print the treatment group comparisons the user has chosen to analyse
base::cat(base::paste0("- ", config$contrasts, "\n"))
```

```{r, results = "asis"}
# print the number of samples analysed
base::cat(base::paste0("Total number of samples: ", base::length(base::unique(metadata$sample))))
```

Number of samples in each treatment group:

```{r, results = "asis"}
# print the number of samples in each treatment group
n_samples_by_treatment <- metadata %>%
  group_by(treatment) %>%
  dplyr::summarise(n_samples = n())

base::cat(base::paste0("- ", n_samples_by_treatment$treatment, ": ", n_samples_by_treatment$n_samples, "\n"))
```

```{r, results = "hide"}
# create a list defining which code chunks to analyse (that are set to TRUE for both the datasets and contrasts to analyse) based on the yaml user configuration file
to_eval_chunk <- config[c("transcript_rnaseq",
                          "gene_rnaseq")] %>%
  rep(times = 10)

names(to_eval_chunk)[1:2] <- paste(names(to_eval_chunk)[1:2], "1", sep = "_")
names(to_eval_chunk)[3:4] <- paste(names(to_eval_chunk)[3:4], "2", sep = "_")
names(to_eval_chunk)[5:6] <- paste(names(to_eval_chunk)[5:6], "3", sep = "_")
names(to_eval_chunk)[7:8] <- paste(names(to_eval_chunk)[7:8], "4", sep = "_")
names(to_eval_chunk)[9:10] <- paste(names(to_eval_chunk)[9:10], "5", sep = "_")
names(to_eval_chunk)[11:12] <- paste(names(to_eval_chunk)[11:12], "6", sep = "_")
names(to_eval_chunk)[13:14] <- paste(names(to_eval_chunk)[13:14], "7", sep = "_")
names(to_eval_chunk)[15:16] <- paste(names(to_eval_chunk)[15:16], "8", sep = "_")
names(to_eval_chunk)[17:18] <- paste(names(to_eval_chunk)[17:18], "9", sep = "_")
names(to_eval_chunk)[19:20] <- paste(names(to_eval_chunk)[19:20], "10", sep = "_")

n_contrasts <- base::length(config$contrasts)

to_eval_chunk <- if(n_contrasts == 1) {
  replace(to_eval_chunk, 3:20, FALSE)
} else if(n_contrasts == 2) {
  replace(to_eval_chunk, 5:20, FALSE)
} else if(n_contrasts == 3) {
  replace(to_eval_chunk, 7:20, FALSE)
} else if(n_contrasts == 4) {
  replace(to_eval_chunk, 9:20, FALSE)
} else if(n_contrasts == 5) {
  replace(to_eval_chunk, 11:20, FALSE)
} else if(n_contrasts == 6) {
  replace(to_eval_chunk, 13:20, FALSE)
} else if(n_contrasts == 7) {
  replace(to_eval_chunk, 15:20, FALSE)
} else if(n_contrasts == 8) {
  replace(to_eval_chunk, 17:20, FALSE)
} else if(n_contrasts == 9) {
  replace(to_eval_chunk, 19:20, FALSE)
} else if(n_contrasts == 10) {
  to_eval_chunk
}

# create a list defining which code chunks to analyse (based on the number of contrasts to analyse) based on the yaml user configuration file
contrasts_chunk_eval <- list(contrast_1 = TRUE,
                             contrast_2 = TRUE,
                             contrast_3 = TRUE,
                             contrast_4 = TRUE,
                             contrast_5 = TRUE,
                             contrast_6 = TRUE,
                             contrast_7 = TRUE,
                             contrast_8 = TRUE,
                             contrast_9 = TRUE,
                             contrast_10 = TRUE)

contrasts_chunk_eval <- if(n_contrasts == 1) {
  replace(contrasts_chunk_eval, 2:10, FALSE)
} else if(n_contrasts == 2) {
  replace(contrasts_chunk_eval, 3:10, FALSE)
} else if(n_contrasts == 3) {
  replace(contrasts_chunk_eval, 4:10, FALSE)
} else if(n_contrasts == 4) {
  replace(contrasts_chunk_eval, 5:10, FALSE)
} else if(n_contrasts == 5) {
  replace(contrasts_chunk_eval, 6:10, FALSE)
} else if(n_contrasts == 6) {
  replace(contrasts_chunk_eval, 7:10, FALSE)
} else if(n_contrasts == 7) {
  replace(contrasts_chunk_eval, 8:10, FALSE)
} else if(n_contrasts == 8) {
  replace(contrasts_chunk_eval, 9:10, FALSE)
} else if(n_contrasts == 9) {
  replace(contrasts_chunk_eval, 10, FALSE)
} else if(n_contrasts == 10) {
  contrasts_chunk_eval
}

# evalutate/setup minimum logFC threshold
min_logfc <- base::eval(base::parse(text = config$min_lfc))
```

## Results

```{r, results = "asis"}
base::cat(base::paste0("Genes/transcripts with a **p-value** smaller than 0.1 and a log fold change greater/smaller than +-", config$min_lfc, " for the [deseq2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8) differential expression method are reported here."))
```

```{r, results = "asis"}
base::cat(base::paste0(" - 'significant_1%' means the gene/transcript meets the logFC threshold and has an **adjusted p-value** smaller than 0.01"))
```

```{r, results = "asis"}
base::cat(base::paste0(" - 'significant_5%' means the gene/transcript meets the logFC threshold and has an **adjusted p-value** smaller than 0.05"))
```

```{r, results = "asis"}
base::cat(base::paste0(" - 'significant_10%' means the gene/transcript meets the logFC threshold and has an **adjusted p-value** smaller than 0.10"))
```

```{r, results = "asis"}
base::cat(base::paste0(" - 'greater_than_10%' means the gene/transcript meets the logFC threshold and has an **adjusted p-value** smaller than 0.10"))
```

```{r, out.width = "100%"}
# create datatable to explore the data
DT::datatable(deseq_diff_expr_results %>%
                dplyr::filter(p_value <= 0.1) %>%
                dplyr::select(gene_transcript,
                              dataset,
                              pipeline,
                              diff_expr_method,
                              comparison,
                              log_fc,
                              p_value,
                              adj_p_value,
                              significance) %>%
                dplyr::mutate(across(c(dataset,
                                       pipeline,
                                       diff_expr_method,
                                       comparison,
                                       significance), base::as.factor)) %>%
                dplyr::mutate(across(c(log_fc,
                                       p_value,
                                       adj_p_value), base::as.double)) %>%
                dplyr::mutate(across(c(p_value,
                                       adj_p_value), ~base::round(.x, digits = 10))) %>%
                dplyr::mutate(across(c(log_fc), ~base::round(.x, digits = 4))),
              filter = "top",
              rownames = FALSE,
              colnames = c("Gene/transcript",
                           "Dataset",
                           "Pipeline",
                           "Differential expression method",
                           "Comparison",
                           "Log fold change",
                           "p-value",
                           "Adjusted p-value",
                           "Significance"),
              extensions = base::list("ColReorder" = NULL,
                                      "Buttons" = NULL,
                                      "FixedColumns" = base::list(leftColumns=1)),
              options = list(
                dom = "BRrltpi",
                autoWidth = TRUE,
                buttons =
                  list("copy",
                       list(extend = "collection",
                            buttons = c("csv", "excel", "pdf"),
                            text = "Download"),
                       I("colvis"))))
```

```{r, cleanup, results = "hide"}
# clean up
rm(list = ls())
```
